rotkehlchen/chain/ethereum/modules/makerdao/sai/decoder.py

```diff
@@ -0,0 +1,505 @@
+MAKERDAO_SAITAP_CONTRACT = string_to_evm_address('0xBda109309f9FafA6Dd6A9CB9f1Df4085B27Ee8eF')
+MAKERDAO_SAI_PROXY_CONTRACT = string_to_evm_address('0x526af336D614adE5cc252A407062B8861aF998F5')
+TOPICS_TO_METHODS = {
+    b'\x89\xb8\x89;\x80m\xb5\x08\x97\xc8\xe26,qW\x1c\xfa\xeb\x97a\xee@r\x7fh?\x17\x93\xcd\xa9\xdf\x16': '_decode_new_cdp_event',  # noqa: E501
+    b'\xb8M!\x06\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': '_decode_close_cdp',  # noqa: E501
+    b'D\x0f\x19\xba\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': '_decode_borrow_sai_event',  # noqa: E501
+    b's\xb3\x81\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': '_decode_repay_sai_event',  # noqa: E501
+    b'\x04\x98x\xf3\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': '_decode_deposit_weth_eth_event',  # noqa: E501
+    b'@\xcc\x88T\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': '_decode_liquidate_event',  # noqa: E501
+    b'\xa5\xcd\x18N\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00': '_decode_withdraw_eth',  # noqa: E501
+}
+
+
+class MakerdaosaiDecoder(DecoderInterface):
+    """
+    Information relating to event topics and contracts came from:
+    https://github.com/makerdao/sai/blob/d28fe5074af7c661fb3aa7f117ad419fb6414b91/DEVELOPING.md
+    """
+    def __init__(
+            self,
+            ethereum_inquirer: 'EthereumInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(ethereum_inquirer, base_tools, msg_aggregator)
+        self.ethereum_inquirer = ethereum_inquirer
+        self.base_tools = base_tools
+        self.weth = A_WETH.resolve_to_evm_token()
+        self.eth = A_ETH.resolve_to_crypto_asset()
+        self.sai = A_SAI.resolve_to_evm_token()
+        # NotERC20Conformant is never raised because PETH is a valid ERC20 token
+        self.peth = get_or_create_evm_token(
+            userdb=self.base_tools.database,
+            chain_id=ChainID.ETHEREUM,
+            token_kind=EvmTokenKind.ERC20,
+            evm_address=POOLED_ETHER_ADDRESS,
+            evm_inquirer=self.ethereum_inquirer,
+        )
+
+    def _decode_sai_tub(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],
+            action_items: Optional[list[ActionItem]],
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        try:
+            decoder_function = getattr(self, TOPICS_TO_METHODS[tx_log.topics[0]])
```

Comment:
dude no .. what are you doing? This won't work. This is not an attribute of self, you have it declared in the module level, which is also wrong. Wtf?

What you need to do is:
1. In the constructor do `self.topics_to_methods = {bla bla}.
2. In every method get it by `self.topics_to_methods[tx_log.topics[0]]`